File: DSA/YourDostAssignment.js
function findSecondLargestUnique(arr) {
    if (arr.length < 2) return -1;

    arr.sort((a, b) => a - b);
    let largest = arr[arr.length - 1];

    let low = 0;
    let high = arr.length - 1;
    let result = -1;

    // binary search for the largest value < largest
    while (low <= high) {
        let mid = Math.floor(low + (high - low) / 2);

        if (arr[mid] < largest) {
            result = arr[mid];
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return result;
}

// sample input (replace with input logic if needed)
const arr = [4, 2, 9, 9, 9, 7];
const result = findSecondLargestUnique(arr);

if (result === -1) {
    console.log(result);
    console.log("Not enough unique elements");
} else {
    console.log("second largest unique element is:", result);
}

File: DSA/YourDostAssignment.py
def find_second_largest_unique(arr):
    if len(arr) < 2:
        return -1

    arr.sort()
    largest = arr[-1]

    low = 0
    high = len(arr) - 1
    result = -1

    # binary search for the largest value < largest
    while low <= high:
        mid = low + (high - low) // 2

        if arr[mid] < largest:
            result = arr[mid]
            low = mid + 1
        else:
            high = mid - 1

    return result


# main
n = int(input())
arr = list(map(int, input().split()))

res = find_second_largest_unique(arr)

if res == -1:
    print(res)
    print("Not enough unique elements")
else:
    print("second largest unique element is:", res)
